core/server/master/src/main/java/alluxio/master/file/meta/InodeTree.java

```diff
@@ -590,56 +648,46 @@ public CreatePathResult createPath(RpcContext rpcContext, LockedInodePath inodeP
-        dir = InodeDirectory.create(
+        InodeDirectory newDir = InodeDirectory.create(
             mDirectoryIdGenerator.getNewDirectoryId(rpcContext.getJournalContext()),
             currentInodeDirectory.getId(), pathComponents[k], missingDirOptions);
         // Lock the newly created inode before subsequent operations, and add it to the lock group.
-        extensibleInodePath.getLockList().lockWriteAndCheckNameAndParent(dir,
+        extensibleInodePath.getLockList().lockWriteAndCheckNameAndParent(newDir,
             currentInodeDirectory, pathComponents[k]);
 
-        if (!currentInodeDirectory.addChild(dir)) {
-          // The child directory inode already exists. Get the existing child inode.
-          extensibleInodePath.getLockList().unlockLast();
+        newDir.setPinned(currentInodeDirectory.isPinned());
 
-          dir =
-              (InodeDirectory) currentInodeDirectory.getChildReadLock(pathComponents[k],
-                  extensibleInodePath.getLockList());
-          if (dir == null) {
-            // Could not get the child inode. Continue and try again.
-            continue;
-          }
-        } else {
-          try {
-            // Successfully added the child, while holding the write lock.
-            dir.setPinned(currentInodeDirectory.isPinned());
-
-            // if the parent has default ACL, copy that default ACL as the new directory's default
-            // and access acl.
-            if (!options.isMetadataLoad()) {
-              DefaultAccessControlList dAcl = currentInodeDirectory.getDefaultACL();
-              if (!dAcl.isEmpty()) {
-                Pair<AccessControlList, DefaultAccessControlList> pair = dAcl.generateChildDirACL();
-                dir.setInternalAcl(pair.getFirst());
-                dir.setDefaultACL(pair.getSecond());
-              }
-            }
-            if (options.isPersisted()) {
-              // Do not journal the persist entry, since a creation entry will be journaled instead.
-              syncPersistDirectory(RpcContext.NOOP, dir);
-            }
-          } catch (Exception e) {
-            // Failed to persist the directory, so remove it from the parent.
-            currentInodeDirectory.removeChild(dir);
-            throw e;
+        // if the parent has default ACL, copy that default ACL as the new directory's default
+        // and access acl.
+        if (!options.isMetadataLoad()) {
+          DefaultAccessControlList dAcl = currentInodeDirectory.getDefaultACL();
+          if (!dAcl.isEmpty()) {
+            Pair<AccessControlList, DefaultAccessControlList> pair = dAcl.generateChildDirACL();
+            newDir.setInternalAcl(pair.getFirst());
+            newDir.setDefaultACL(pair.getSecond());
           }
-          // Journal the new inode.
-          rpcContext.getJournalContext().append(dir.toJournalEntry());
-          mInodes.add(dir);
+        }
+        if (options.isPersisted()) {
+          syncPersistNewDirectory(newDir);
```

Comment:
I think this ordering is different from before?

Reply:
Moved the `syncPersist` to be after the directory is added to the tree.

I noticed that previously we would only call `syncPersistDirectory` in the case where we successfully add our new directory to the tree. Otherwise we take the existing directory and continue. But what if the pre-existing directory is not persisted, but our `options.isPersisted()` is true? I think we should call `syncPersistDirectory` whether we add a new directory, or discover an existing one with the same name. What do you think?

Reply:
wow, great catch. I think you are right. We should `syncPersist` the inode (if `isPersisted`) for both of those scenarios.

Now that I think about it, I think the previous order might be ok too, because `syncPersist` should serialize the UFS create with its own locking behavior?